client/packages/core/src/queryValidation.ts

```diff
@@ -0,0 +1,380 @@
+import { IContainEntitiesAndLinks, DataAttrDef } from './schemaTypes.ts';
+
+export class QueryValidationError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'QueryValidationError';
+  }
+}
+
+const dollarSignKeys = [
+  'where',
+  'order',
+  'limit',
+  'last',
+  'first',
+  'offset',
+  'after',
+  'before',
+  'fields',
+  'aggregate',
+];
+
+const getAttrType = (attrDef: DataAttrDef<any, any, any>): string => {
+  if (attrDef.valueType === 'string') return 'string';
+  if (attrDef.valueType === 'number') return 'number';
+  if (attrDef.valueType === 'boolean') return 'boolean';
+  if (attrDef.valueType === 'date') return 'date';
+  if (attrDef.valueType === 'json') return 'object';
+  return 'unknown';
+};
+
+const isValidValueForType = (
+  value: unknown,
+  expectedType: string,
+  isAnyType: boolean = false,
+): boolean => {
+  if (isAnyType) return true;
+  if (value === null || value === undefined) return true;
+
+  switch (expectedType) {
+    case 'string':
+      return typeof value === 'string';
+    case 'number':
+      return typeof value === 'number' && !isNaN(value);
+    case 'boolean':
+      return typeof value === 'boolean';
+    case 'date':
+      return value instanceof Date || typeof value === 'string';
+    case 'object':
+      return typeof value === 'object' && value !== null;
+    default:
+      return true;
+  }
+};
+
+const validateWhereClauseValue = (
+  value: unknown,
+  attrName: string,
+  attrDef: DataAttrDef<any, any, any>,
+  entityName: string,
+): void => {
+  const expectedType = getAttrType(attrDef);
+  const isAnyType = attrDef.valueType === 'json';
+
+  if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
+    // For any type, allow complex objects without treating them as operators
+    if (isAnyType) {
+      return; // Any type accepts any value, including complex objects
+    }
+
+    const operators = value as Record<string, unknown>;
+
+    for (const [op, opValue] of Object.entries(operators)) {
+      switch (op) {
```

Comment:
It may make sense to extract this bit into a helper function, to make reading this part of the function a bit easier